# First Homework: Linear and Discrete Models 

## Marta Almagro Fuello    
#### NIA : 100451979


## Problem statement 
### a) Formulate the problem as a linear optimization model (general formulation). Identify the model sets, parameters, variables, objective function and constraints.

Is very common that during the last year of every degree most of the students that are keeping up with them grades start searching for interships in companies to put in practice what they have learned in their subjects of college.

At the same time, big companies are also searching for this type of students so they can help in proyects even though they do not have experience, they can contribute in them. Another reason for the companies to constract this non-graduate students is because they are much cheaper to pay than a person who actually has a degree.

I have selected the "Big 4" companies plus Accenture, which is a big firm too.

<img src = "img/BIG4.jpg" alt = "Companies" />

Each university has made different salary agreements with each company for the interships that are reflected in the below table

<img src="img/tablepart1.jpg" alt = "TablePart1" />



In the table below it is represented the number of jobs offered from each firm and the minimum number of positions that need to be covered. In addition there is reflected also the maximum number of students that each univeristy can allow and the minimum number or agreements they have to sign so these univeristies get in the top 10 of "Best Universities" from Madrid.

The government wants to sign limited agreements between the universities and companies with the least possible total cost. One of the conditons of the contract is that it is compulsory for each company to hire between 2 and 25 students from each University.

They want to be fair with private and public universities, so it does not seems that companies only hire students from the public universities. In order to acomplish that, they stated that
at least half the students that get interships should be studying in a private univeristy.
The private universities are CEU, ICADE and ESIC, the rest are public.



The Intership Problem can be formulated mathematically as a linear programming problem using the following model.


### Sets

 $C$ = set of Companies
 
 $U$ = set of Universities
 
 
### Parameters

$StudentsAvailable$ = students offered by each univeristy $i$,$\forall j\in C$

$JobsNeededCovered$ = students demanded by each company $j$, $\forall i\in U$

$MinAgreementsbyUni$ = minimum number of scholars that each university $i$ must offer,$\forall j\in C$

$JobsOfferedbyCompany$ = maximum number of scholars that each company $j$, is able to contract, $\forall i\in U$

$Salaries$ = $salary_{ij}$ paid per each scholar

### Variables

 $x_{ij}$ = number of students hired from each university $i$ to each company $j$

### Objective

Minimize the total cost of scholars salaries  

 $\min \sum_{i\in U}\sum_{j \in C} salaries_{ij} * x_{ij}$

### Constraints

Limited number of students that are able to work in each university 

$\sum_{i \in U} x_{ij} \leq StudentsAvailable_i $, $\forall j \in C$

Minimum number of scholars required in each company

$\sum_{j \in C} x_{ij} \geq JobsNeededCovered_j $, $\forall i \in U$


 
Minimum number of students each university sent to each company

$\sum_{i \in U} x_{ij} \geq MinAgreementsbyUni_i $, $\forall j \in C$
 
Maximum number of scholars each company can accept

$\sum_{j \in C} x_{ij} \leq JobsOfferedbyCompany_j $, $\forall i \in U$

At least half the students that get interships should be studying in a private univeristy

$\sum_{j \in C} x_{ij} \geq 0.5 * (CEU_i +  ICADE_i + ESIC_i) $, $\forall i \in U$



### b) Implement the model in a notebook (preferably using Pyomo, but not mandatory) and solve it for an input based on real (or realistic) world data. Interpret the solution.

First we must import the libraries we are going to use that are:

In [1]:
from pyomo.environ import *
from pyomo.opt import SolverFactory

Now we can create the model and start declaring the data:

In [2]:
# Creating the model

model = ConcreteModel('Intership Problem')
model.dual = Suffix(direction = Suffix.IMPORT)

# Defining the sets

model.i = Set(initialize = ['UC3M','UCM','URJC','CEU','ICADE','ESIC'], doc = 'Universities' ) 
model.j = Set(initialize = ['ACCENTURE','KPMG','EY','PWC','DELOITTE'], doc = 'Companies') 

## Defining values for the sets

# Define the students available by each university and the jobs offered by each company
StudentsAvailable = {'UC3M': 79, 'UCM': 65, 'URJC': 63, 'CEU':61, 'ICADE':42, 'ESIC':58}
JobsOfferedbyCompany = {'ACCENTURE': 83, 'KPMG': 79, 'EY': 77, 'PWC':91,'DELOITTE':85}

model.StudentsAvailable = Param(model.i, initialize = StudentsAvailable) 
model.JobsOffered = Param(model.j, initialize = JobsOfferedbyCompany)

# Define the minimum of agreements signed by each university 
# and the jobs that have to be covered by each company
MinAgreementsbyUni = {'UC3M': 15, 'UCM': 20, 'URJC': 12, 'CEU':58, 'ICADE':15, 'ESIC':14}
JobsNeededCovered = {'ACCENTURE': 30, 'KPMG': 33, 'EY': 28, 'PWC':46,'DELOITTE':24}

model.MinAgreementsbyUni = Param(model.i, initialize = MinAgreementsbyUni)
model.JobsNeededCovered = Param(model.j, initialize = JobsNeededCovered) 

# Define the salaries

Salaries = {
    ('UC3M', 'ACCENTURE' ): 1950,
    ('UC3M', 'KPMG'): 2100,
    ('UC3M', 'EY' ): 1200,
    ('UC3M', 'PWC' ): 1500,
    ('UC3M', 'DELOITTE' ): 1200,
    ('UCM', 'ACCENTURE'): 2010,
    ('UCM', 'KPMG'): 1850,
    ('UCM', 'EY'): 1800,
    ('UCM', 'PWC'): 1750,
    ('UCM', 'DELOITTE' ): 1825,    
    ('URJC', 'ACCENTURE'): 1200,
    ('URJC', 'KPMG'): 1200,
    ('URJC', 'EY'): 1500,
    ('URJC', 'PWC'): 1200,
    ('URJC', 'DELOITTE' ): 2000,
    ('CEU', 'ACCENTURE'): 1200,
    ('CEU', 'KPMG'): 1800,
    ('CEU', 'EY'): 1700,
    ('CEU', 'PWC'): 1000,
    ('CEU', 'DELOITTE' ): 1500,
    ('ICADE', 'ACCENTURE'): 1500,
    ('ICADE', 'KPMG'): 2000,
    ('ICADE', 'EY'): 1560,
    ('ICADE', 'PWC'): 1810,
    ('ICADE', 'DELOITTE' ): 1800,    
    ('ESIC', 'ACCENTURE' ): 1800,
    ('ESIC', 'KPMG' ): 1500,
    ('ESIC', 'EY' ): 1350,
    ('ESIC', 'PWC' ): 1350,
    ('ESIC', 'DELOITTE' ): 1200,
    } 

model.Salaries = Param(model.i, model.j, initialize = Salaries)
#model.pprint()

Secondly, the decision variable is created taking into account the goverment's condition:

In [3]:
model.x = Var(model.i, model.j, within = NonNegativeReals, bounds = (2,25))

Once we have all then parameters and variables created we can start declaring the objective and constraint functions:

In [4]:
def objective_rule(model):
    return sum(model.Salaries[i,j] * model.x[i,j] for i in model.i for j in model.j)

model.objective = Objective(rule = objective_rule, sense = minimize)
model.objective.pprint()

objective : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : minimize : 1950*x[UC3M,ACCENTURE] + 2100*x[UC3M,KPMG] + 1200*x[UC3M,EY] + 1500*x[UC3M,PWC] + 1200*x[UC3M,DELOITTE] + 2010*x[UCM,ACCENTURE] + 1850*x[UCM,KPMG] + 1800*x[UCM,EY] + 1750*x[UCM,PWC] + 1825*x[UCM,DELOITTE] + 1200*x[URJC,ACCENTURE] + 1200*x[URJC,KPMG] + 1500*x[URJC,EY] + 1200*x[URJC,PWC] + 2000*x[URJC,DELOITTE] + 1200*x[CEU,ACCENTURE] + 1800*x[CEU,KPMG] + 1700*x[CEU,EY] + 1000*x[CEU,PWC] + 1500*x[CEU,DELOITTE] + 1500*x[ICADE,ACCENTURE] + 2000*x[ICADE,KPMG] + 1560*x[ICADE,EY] + 1810*x[ICADE,PWC] + 1800*x[ICADE,DELOITTE] + 1800*x[ESIC,ACCENTURE] + 1500*x[ESIC,KPMG] + 1350*x[ESIC,EY] + 1350*x[ESIC,PWC] + 1200*x[ESIC,DELOITTE]


In [5]:
# Define the constraint that says the maximum students available by each university  
def StudentsAvailable_rule(model, i):
    return sum(model.x[i,j] for j in model.j) <= model.StudentsAvailable[i]

model.StudentsAvailableCons = Constraint(model.i, rule = StudentsAvailable_rule)
   
# Define the constraint that says the maximum jobs offered by each company   
def JobsOffered_rule(model, j):
    return sum(model.x[i,j] for i in model.i) <= model.JobsOffered[j]

model.JobsOfferedCons = Constraint(model.j, rule = JobsOffered_rule)

# Define the constraint that says the minimum of agreements signed by each university 
def MinAgreementsbyUni_rule(model, i):
    return sum(model.x[i,j] for j in model.j) >= model.MinAgreementsbyUni[i]

model.MinAgreementsbyUniCons = Constraint(model.i, rule = MinAgreementsbyUni_rule)

# Define the constraint that says the minimum jobs that have to be covered by each company
def JobsNeededCovered_rule(model, j):
    return sum(model.x[i,j] for i in model.i) >= model.JobsNeededCovered[j]

model.JobsNeededCoveredCons = Constraint(model.j, rule = JobsNeededCovered_rule)

#Declaring the constraint stated by the goverment to be fair with public and private universities
def PrivateUniveristy_rule(model, j):
    return model.x['UC3M', j] + model.x['UCM', j] + model.x['URJC', j] + model.x['ESIC', j] + model.x['CEU', j] + model.x['ICADE', j] >= 0.5 * (model.x['ICADE', j] + model.x['ESIC', j] + model.x['CEU', j])

model.PrivateUniveristyCons = Constraint(model.j, rule = PrivateUniveristy_rule)

Finally the optimization problem is resolve with function .solve from the package 'glpk'

In [6]:
# Define a solver 
Solver = SolverFactory('glpk')

# Obtain the solution
Results = Solver.solve(model)

# Display the solution
model.x.display()

x : Size=30, Index=x_index
    Key                    : Lower : Value : Upper : Fixed : Stale : Domain
      ('CEU', 'ACCENTURE') :     2 :  20.0 :    25 : False : False : NonNegativeReals
       ('CEU', 'DELOITTE') :     2 :   9.0 :    25 : False : False : NonNegativeReals
             ('CEU', 'EY') :     2 :   2.0 :    25 : False : False : NonNegativeReals
           ('CEU', 'KPMG') :     2 :   2.0 :    25 : False : False : NonNegativeReals
            ('CEU', 'PWC') :     2 :  25.0 :    25 : False : False : NonNegativeReals
     ('ESIC', 'ACCENTURE') :     2 :   2.0 :    25 : False : False : NonNegativeReals
      ('ESIC', 'DELOITTE') :     2 :   6.0 :    25 : False : False : NonNegativeReals
            ('ESIC', 'EY') :     2 :   2.0 :    25 : False : False : NonNegativeReals
          ('ESIC', 'KPMG') :     2 :   2.0 :    25 : False : False : NonNegativeReals
           ('ESIC', 'PWC') :     2 :   2.0 :    25 : False : False : NonNegativeReals
    ('ICADE', 'ACCENTURE') :     2 : 

## Conclusions and results

In [7]:
print('Minimum budget for a year:',12*int(model.objective()), 'euros')
# Minimum budget for a year: 2.648.520 euros

Minimum budget for a year: 2648520 euros


In [8]:
# Fancy model.x.display() 
for i in model.i:
    for j in model.j:
            print('There are',int(model.x[i,j].value),'students working at',j,'from the univeristy',i+".")

There are 2 students working at ACCENTURE from the univeristy UC3M.
There are 2 students working at KPMG from the univeristy UC3M.
There are 13 students working at EY from the univeristy UC3M.
There are 2 students working at PWC from the univeristy UC3M.
There are 3 students working at DELOITTE from the univeristy UC3M.
There are 2 students working at ACCENTURE from the univeristy UCM.
There are 2 students working at KPMG from the univeristy UCM.
There are 2 students working at EY from the univeristy UCM.
There are 12 students working at PWC from the univeristy UCM.
There are 2 students working at DELOITTE from the univeristy UCM.
There are 2 students working at ACCENTURE from the univeristy URJC.
There are 23 students working at KPMG from the univeristy URJC.
There are 2 students working at EY from the univeristy URJC.
There are 3 students working at PWC from the univeristy URJC.
There are 2 students working at DELOITTE from the univeristy URJC.
There are 20 students working at ACCENT

Minimum budget for a year is 2.648.520 euros with the number of students shown above

## Sensivity

### c) Compute the sensitivities associated with each constraint, and interpret their values.

In [9]:
display(model.dual)

dual : Direction=Suffix.IMPORT, Datatype=Suffix.FLOAT
    Key                              : Value
    JobsNeededCoveredCons[ACCENTURE] :  900.0
     JobsNeededCoveredCons[DELOITTE] : 1200.0
           JobsNeededCoveredCons[EY] : 1200.0
         JobsNeededCoveredCons[KPMG] : 1200.0
          JobsNeededCoveredCons[PWC] : 1200.0
          JobsOfferedCons[ACCENTURE] :    0.0
           JobsOfferedCons[DELOITTE] :    0.0
                 JobsOfferedCons[EY] :    0.0
               JobsOfferedCons[KPMG] :    0.0
                JobsOfferedCons[PWC] :    0.0
         MinAgreementsbyUniCons[CEU] :  300.0
        MinAgreementsbyUniCons[ESIC] :    0.0
       MinAgreementsbyUniCons[ICADE] :  360.0
        MinAgreementsbyUniCons[UC3M] :    0.0
         MinAgreementsbyUniCons[UCM] :  550.0
        MinAgreementsbyUniCons[URJC] :    0.0
    PrivateUniveristyCons[ACCENTURE] :    0.0
     PrivateUniveristyCons[DELOITTE] :    0.0
           PrivateUniveristyCons[EY] :    0.0
         PrivateUniveristyC

After this analysis we can verify that every contraint that has been stated is taken into account.
However there are two constraints that have all the influence of the problem which are MinAgreementsbyUniCons and JobsNeededCoveredCons, I tried to change the data so the rest have influence and it is a more realistic problem but the budget cost increased.

# The binary intership problem

### d) Modify the problem in a) to impose some logical and conditional constraints that require the use of binary or integer variables. The more integer variables or constraints, the better. Implement and solve this new model and interpret the results.

## Summary

In the previous problem we where calculating the number of students per company depending on the salaries, because we wanted the minimun budget, in this part we are going to choose them depending on the efficiency of the students, and the form of measuring this quality is taking into account the number of project each scholar is able to do in a year.


<img src = "img/download.jpg" alt = "Companies" />

 
In addition, due to the economic crisis, the budget is lower and as a consequence we can afford a lower number of students per company. Which means that in this case each company offers less places for the scholars.

As the perfils of some univeristies are very different and work distint ways, the companies do not want them working  together because they are no compatible. The students of the universities that cannot work together are "UC3M with ICADE" , "UCM with CEU" and "URJC and ESIC". However there are some, that in the contrary, are very similar and have the same methods implied in their practices that make them a good team, these are: "UCM and ICADE" and "ICADE with ESIC".

The data:

<img src = "img/tablepart2.jpg" alt = "Table2" />


In this case I have decided to generate the data in excel in order to define the variables and access to the data faster, and also to practice different ways of working with the data within the same assignment.

## New problem statement

The Intership Binary Problem can be formulated mathematically as a linear programming problem using the following model.

 
### Parameters

NumStudents = number of students $ij$, $\forall j\in C$ ,  $\forall i\in U$

ProjectsDone = Number of expected projects done $i$, $\forall j\in C$

MaxStudents = Maximum number of students $j$, $\forall i\in U$

MinStudent = 1 

### Variables

$$
x_{ij}=
\left\{\begin{array}{ll} 
1, & \text{if the university $i$ sends their students to companies $j$,}\\
0, & \text{if the university $i$ does not send their students to companies $j$}
\end{array} \right.\quad
$$

### Objective

Maximize the expected projects done

$\max \sum_{i\in U}\sum_{j \in C} ProjectsDone_{i}*x_{ij}$

### Constraints

Maximum Number of Students:

$\sum_{i \in U} x_{ij} \leq MaxStudents_i $, $\forall j \in C$

Minimum Number of Students:

$\sum_{i \in U} x_{ij} \geq MinStudents $, $\forall j \in C$

#### Logical constraints

$\sum_{i \in 1,5} x_{i} \leq 1 $, $\forall j \in C$

$\sum_{i \in 6,3} x_{i} \leq 1 $, $\forall j \in C$

$\sum_{i \in 2,4} x_{i} \leq 1 $, $\forall j \in C$

$\sum_{i \in 2,5} x_{i} \geq 0 $, $\forall j \in C$

$\sum_{i \in 6,5} x_{i} \geq 0 $, $\forall j \in C$

Firstly we must import the dataset and declare which columns we are going to use:

In [10]:
import pandas as pd

data = pd.read_excel('Workbook1.xlsx', usecols = 'A:G') 
print(data)

   Universities  ACCENTURE  KPMG  EY  PWC  DELOITTE  PROJECTS DONE
0          UC3M         15     0   9    7        12          200.0
1           UCM          3     3   6    0         2          125.0
2          URJC         16     4   0    3         3           95.0
3           CEU          6    12   3    0         4          175.0
4         ICADE         13     0   0   11         5          210.0
5          ESIC          0    10   8   15         7          140.0
6  JOBS OFFERED         30    22  20   25        20            NaN


Once the data has been correctly imported we can define each of the parameters:

In [11]:
Universities = data.iloc[0:6,0]
Companies = data.columns[1:6]
ProjectsDone = data.iloc[0:6,6]
NumStudents = data.iloc[0:6,1:6]
MinStudents = 1
MaxStudents = data.iloc[6,1:6]

Start defining the binary variables and objective rule:

In [12]:
model1 = ConcreteModel('Intership Binary Problem')

# Index for universities and companies
model1.Universities = RangeSet(1, len(Universities)) # [1:6]
model1.Companies = RangeSet(1, len(Companies)) # [1:5]

# Binary variables
model1.x = Var(model1.Universities, domain = Binary)

# Objective function
def objective_rule(model1):
    return sum(ProjectsDone[i-1] * model1.x[i] for i in model1.Universities)

model1.objective = Objective(rule = objective_rule, sense = maximize)

# As we are using indexes in some variables they count the zero so that is why
# we are substracting 1 in the ones that do not start counting by 0

In [13]:
print("The objective expression is going to be:")
model1.objective.pprint()

The objective expression is going to be:
objective : Size=1, Index=None, Active=True
    Key  : Active : Sense    : Expression
    None :   True : maximize : 200.0*x[1] + 125.0*x[2] + 95.0*x[3] + 175.0*x[4] + 210.0*x[5] + 140.0*x[6]


In [14]:
# Constraints:

def max_students(model1, j): 
    return sum(NumStudents.values[i-1, j-1] * model1.x[i] for i in model1.Universities) <= MaxStudents.values[j-1]

model1.MaxStudents = Constraint(model1.Companies, rule = max_students)

def min_students(model1, j): 
    return sum(NumStudents.values[i-1, j-1] * model1.x[i] for i in model1.Universities) >= MinStudents

model1.MinStudents = Constraint(model1.Companies, rule = min_students)

In [15]:
# Solution
# Define the solver
opt = SolverFactory('glpk')

In [16]:
# Logical Constraints:

model1.log1 = Constraint(expr = model1.x[1] + model1.x[5] <= 1) # UC3M with ICADE
model1.log2 = Constraint(expr = model1.x[6] + model1.x[3] <= 1) # UCM with CEU
model1.log3 = Constraint(expr = model1.x[2] + model1.x[4] <= 1) # URJC and ESIC
model1.log4 = Constraint(expr = model1.x[2] + model1.x[5] <= 0) # UCM and ICADE
model1.log5 = Constraint(expr = model1.x[6] + model1.x[5] <= 0) # ICADE with ESIC

model1.pprint()

2 RangeSet Declarations
    Companies : Dimen=1, Size=5, Bounds=(1, 5)
        Key  : Finite : Members
        None :   True :   [1:5]
    Universities : Dimen=1, Size=6, Bounds=(1, 6)
        Key  : Finite : Members
        None :   True :   [1:6]

1 Var Declarations
    x : Size=6, Index=Universities
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          1 :     0 :  None :     1 : False :  True : Binary
          2 :     0 :  None :     1 : False :  True : Binary
          3 :     0 :  None :     1 : False :  True : Binary
          4 :     0 :  None :     1 : False :  True : Binary
          5 :     0 :  None :     1 : False :  True : Binary
          6 :     0 :  None :     1 : False :  True : Binary

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 200.0*x[1] + 125.0*x[2] + 95.0*x[3] + 175.0*x[4] + 210.0*x[5] + 140.0*x[6]

7 Constraint Declarations
    MaxStude

In [17]:
# Solve
results = opt.solve(model1)  # solve the model with the selected solver
 
model1.display()

model1.objective()

Model Intership Binary Problem

  Variables:
    x : Size=6, Index=Universities
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          1 :     0 :   1.0 :     1 : False : False : Binary
          2 :     0 :   0.0 :     1 : False : False : Binary
          3 :     0 :   0.0 :     1 : False : False : Binary
          4 :     0 :   1.0 :     1 : False : False : Binary
          5 :     0 :   0.0 :     1 : False : False : Binary
          6 :     0 :   0.0 :     1 : False : False : Binary

  Objectives:
    objective : Size=1, Index=None, Active=True
        Key  : Active : Value
        None :   True : 375.0

  Constraints:
    MaxStudents : Size=5
        Key : Lower : Body : Upper
          1 :  None : 21.0 :  30.0
          2 :  None : 12.0 :  22.0
          3 :  None : 12.0 :  20.0
          4 :  None :  7.0 :  25.0
          5 :  None : 16.0 :  20.0
    MinStudents : Size=5
        Key : Lower : Body : Upper
          1 :   1.0 : 21.0 :  None
          2 :   1.0 : 12

375.0

In [18]:
print('Total expected projects done =', model1.objective())
print('\nSelected Universities:')
for i in model1.Universities:
    if model1.x[i]() > 0:
        print('-University', Universities[i-1],'->',model1.x[i]())

Total expected projects done = 375.0

Selected Universities:
-University UC3M -> 1.0
-University CEU -> 1.0


## Conclusion

With the new constraints the target is to maximize the estimated number of projects is finished. 

This is number of total expected projects done is 375 and it is obtanined hiring scholars who are fully compatible from UC3M and CEU universities. 


We can check that this result fulfill all the restrictions:
- First if we sum all the students of each company they are not more than the jobs offered.
- Then we can observe that UC3M and CEU univeristy are not working with ICADE or UCM which are the incompatible universities respectively. Also they do not have the constraint of working with any college.

In [19]:
ACCENTURE = data.iloc[0,1] + data.iloc[3,1]
KPMG = data.iloc[0,2] + data.iloc[3,2] 
EY = data.iloc[0,3] + data.iloc[3,3]
PWC = data.iloc[0,4] + data.iloc[3,4]
DELOITTE = data.iloc[0,5] + data.iloc[3,5]
MaxStudents = data.iloc[6,1:6]
Studentslist = [ACCENTURE, KPMG, EY, PWC, DELOITTE]
acc = 0
for i in range(len(Studentslist)):
    if Studentslist[i] > MaxStudents[i]:
        print("The constraint max_students is not fullfilled")

<img src = "img/uc3m.jpg" alt = "Companies" style="width: 600px; height: 300px;" />